kimu/core/rectangular_tool.py

```diff
@@ -0,0 +1,403 @@
+import math
+from decimal import Decimal
+from typing import List
+
+from qgis.core import (
+    QgsFeature,
+    QgsField,
+    QgsGeometry,
+    QgsPointXY,
+    QgsProject,
+    QgsSnappingConfig,
+    QgsTolerance,
+    QgsVectorLayer,
+    QgsWkbTypes,
+)
+from qgis.gui import QgisInterface, QgsMapToolEmitPoint, QgsSnapIndicator
+from qgis.PyQt.QtCore import QVariant
+from qgis.PyQt.QtGui import QColor
+from qgis.utils import iface
+
+from ..qgis_plugin_tools.tools.custom_logging import setup_logger
+from ..qgis_plugin_tools.tools.i18n import tr
+from ..qgis_plugin_tools.tools.resources import plugin_name
+from ..ui.rectangular_dockwidget import RectangularDockWidget
+from .select_tool import SelectTool
+
+LOGGER = setup_logger(plugin_name())
+
+
+class RectangularMapping(SelectTool):
+    def __init__(
+        self, iface: QgisInterface, dock_widget: RectangularDockWidget
+    ) -> None:
+        super().__init__(iface)
+        self.ui: RectangularDockWidget = dock_widget
+        # Set suitable snapping settings
+        my_snap_config = QgsSnappingConfig()
+        my_snap_config.setEnabled(True)
+        my_snap_config.setType(QgsSnappingConfig.Vertex)
+        my_snap_config.setUnits(QgsTolerance.Pixels)
+        my_snap_config.setTolerance(15)
+        my_snap_config.setIntersectionSnapping(True)
+        my_snap_config.setMode(QgsSnappingConfig.AllLayers)
+        QgsProject.instance().setSnappingConfig(my_snap_config)
+        self.i = QgsSnapIndicator(self.iface.mapCanvas())
+
+    def active_changed(self, layer: QgsVectorLayer) -> None:
+        """Triggered when active layer changes."""
+        if (
+            isinstance(layer, QgsVectorLayer)
+            and layer.isSpatial()
+            and layer.geometryType() == QgsWkbTypes.LineGeometry
+        ):
+            self.layer = layer
+            self.setLayer(self.layer)
+
+    # fmt: off
+    def canvasPressEvent(  # noqa: N802
+        self, event: QgsMapToolEmitPoint
+    ) -> None:
+        # fmt: on
+        """Canvas click event."""
+        if self.iface.activeLayer() != self.layer:
+            LOGGER.warning(tr("Please select a line layer"),
+                           extra={"details": ""})
+            return
+
+        check = 0
+        for test_feature in self.iface.activeLayer().getFeatures():
+            if check == 0:
+                test_geom = test_feature.geometry()
+                if QgsWkbTypes.isSingleType(test_geom.wkbType()):
```

Comment:
If the geometry of the first feature matches this, `check` becomes 1 and all other features are still iterated over but simply passed. Else, `None` is returned. So best to simply check if the geometry of the first feature matches the `isSingleType` condition. This way you can get rid of the unnecessary `check` variable.